### PR TITLE
Notifications + Swift Mk 5: Simperium

### PR DESCRIPTION
**Pull request by jleandroperez**
#### Details:
This PR adds no new functionality. We'll gradually migrate all of the Notifications-Y code over to Swift.
Once it's ready, we'll address code improvements / technical debt.

Reference #5613

--

#### Scenario A: Regular Notification
1. Receive a regular Notification (you may click like over any of your own posts)
2. Launch the app and open the Notifications tab
3. Press over the new notification row

As a result, the notification details should be onscreen.

#### Scenario B: Automatcher
1. Receive a matcher notification (mention on a post)
2. Launch the app and open the Notifications tab
3. Press over the Matcher notification

As a result, the reader should be onscreen.

#### Scenario C: While Onscreen
1. Launch the app and open the Notifications screen
2. Receive a new notification

As a result, the badge should get immediately reset.

#### Scenario D: Background
1. Receive a push notification while the app is in BG
2. Tap over the note to open the details

As a result, the details should become onscreen, after a while

#### Scenario E: Timeout
1. Hack [this spot](https://github.com/wordpress-mobile/WordPress-iOS/blob/issue/notifications-swiftification-mark5/WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController+Extensions.swift#L620) and lower the timeout to, say, 0.1 seconds
2. Receive a push notification while the app is in BG
3. Immediately press on it

As a result, the app should launch, and the notifications list should be onscreen. However, the Details shouldn't show up.

Needs review: @astralbodies 
Aaron, may i bother you with a (hopefully fun) review?

Thanks in advance!

### CODE REVIEW

File: WordPress/Classes/Extensions/NSFetchedResultsController+Helpers.swift [hunk 1]
```diff
@@ -3,6 +3,21 @@ import Foundation
 
 extension NSFetchedResultsController
 {
+    /// Returns whether an indexPath represents the last row in it's section, or not
+    ///
+    func isLastRowInSection(indexPath: NSIndexPath) -> Bool {
```
Comment (astralbodies) [inline WordPress/Classes/Extensions/NSFetchedResultsController+Helpers.swift]: I'd maybe take a section instead of the indexpath just for clarity.

Comment (astralbodies) [inline WordPress/Classes/Extensions/NSFetchedResultsController+Helpers.swift]: Hmm okay you do use the row below. Maybe `forIndexPath` for the parameter? Maybe I'm just being stupid. 😄

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController+Extensions.swift [hunk 2]
```diff
@@ -236,7 +288,8 @@ extension NotificationsViewController: WPTableViewHandlerDelegate
         // after a DB OP. This loop has been measured in the order of milliseconds (iPad Mini)
         for indexPath in tableView.indexPathsForVisibleRows ?? [] {
             let cell = tableView.cellForRowAtIndexPath(indexPath) as! NoteTableViewCell
-            cell.showsBottomSeparator = isRowLastRowForSection(indexPath)
+            let isLastRow = tableViewHandler.resultsController.isLastRowInSection(indexPath)
+            cell.showsBottomSeparator = isLastRow == false
```
Comment (astralbodies) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController+Extensions.swift]: I'd either use `!isLastRow` or enclose this in parenthesis for clarity.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController+Extensions.swift [hunk 3]
```diff
@@ -404,6 +457,90 @@ extension NotificationsViewController
...
+        if pushNotificationID == key {
+            if abs(pushNotificationDate.timeIntervalSinceNow) <= Syncing.pushMaxWait {
+                showDetailsForNotificationWithID(key)
+            }
+
+            stopWaitingForNotification()
+        }
+
+        // Mark as read immediately, if needed
+        if isViewOnScreen() == true && UIApplication.sharedApplication().applicationState == .Active {
```
Comment (astralbodies) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController+Extensions.swift]: This operates as expected when the app is on the iPad in multitasking mode but not in focus, right?

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController+Extensions.swift [hunk 4]
```diff
@@ -404,6 +457,90 @@ extension NotificationsViewController
...
+        pushNotificationDate = nil
+    }
+
+    func notificationWaitDidTimeout() {
+        DDLogSwift.logInfo("Sync Timeout: Cancelling wait for notification with ID [\(pushNotificationID)]")
+
+        pushNotificationID = nil
+        pushNotificationDate = nil
+
+        let properties = [Stats.networkStatusKey : simperium.networkStatus]
```
Comment (astralbodies) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController+Extensions.swift]: knitpick - no space before colon in dictionaries 😀

**Comment by astralbodies** [comment]
I got a crash upon logging back in after logging out as a different user.

File: WordPress/Classes/Extensions/NSFetchedResultsController+Helpers.swift (hunk 1 above)
Comment (jleandroperez) [inline WordPress/Classes/Extensions/NSFetchedResultsController+Helpers.swift]: forIndexPath sounds good, thanks!

Comment (jleandroperez) [inline WordPress/Classes/Extensions/NSFetchedResultsController+Helpers.swift]: Mmmmmm i think i'll just rename it to `isLastIndexPathInSection`. Feels better!

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController+Extensions.swift (hunk 3 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController+Extensions.swift]: Yup! verified on iPad Air 2 + iOS 9. Side by side with Safari (and the focus on Safari!).

**Comment by jleandroperez** [comment]
@astralbodies addressed all of your comments sir. Thanks a lot for catching the IUO bugo!!

Ready for another look!

**Comment by astralbodies** [comment]
This looks good! :shipit:

**Comment by jleandroperez** [comment]
Thanks a lot for the review @astralbodies !!!

**Pull request merged**